R/corPFSOS.R

```diff
@@ -0,0 +1,307 @@
+#' log_p11(transition, 1, 3)
+log_p11 <- function(transition, s, t) {
+  intval <- mapply(function(s, t) stats::integrate(p11Integ, lower = s, upper = t, transition)$value, s, t)
+  -intval
+}
+
+#' Helper Function for `survPFSOS()`
+#'
+#' @param u (`numeric`)\cr variable of integration.
+#' @param t (`numeric`)\cr time at which the value of the PFS*OS survival function is to be computed.
+#' @param transition (`TransitionParameters`)\cr
+#'   see [exponential_transition()], [weibull_transition()] or [piecewise_exponential()] for details.
+#'
+#' @return Numeric result of the integrand used to calculate the PFS*OS survival function.
+#' @export
+#'
+#' @examples
+#' transition <- exponential_transition(h01 = 1.2, h02 = 1.5, h12 = 1.6)
+#' PFSOSInteg(1, 2, transition)
+PFSOSInteg <- function(u, t, transition) {
+  exp(log_p11(transition, u, t / u) + log(survPFS(transition, u)) + log(haz(transition, u, 1)))
+}
+
+#' Survival Function of the Product PFS*OS for Different Transition Models
+#'
+#' @param t (`numeric`)\cr time at which the value of the PFS*OS survival function is to be computed.
+#' @param transition (`TransitionParameters`)\cr
+#'   see [exponential_transition()], [weibull_transition()] or [piecewise_exponential()] for details.
+#'
+#' @return This returns the value of PFS*OS survival function at time t.
+#' @export
+#'
+#' @examples
+#' transition <- exponential_transition(h01 = 1.2, h02 = 1.5, h12 = 1.6)
+#' survPFSOS(0.4, transition)
+survPFSOS <- function(t, transition) {
+  sapply(t, function(x) {
+    intval <- stats::integrate(PFSOSInteg, lower = 0, upper = sqrt(x), x, transition)$value
+    survPFS(transition, sqrt(x)) + intval
+  })
+}
+
+#' Correlation of PFS and OS event times for Different Transition Models
+#'
+#' @param transition (`TransitionParameters`)\cr
+#'   see [exponential_transition()], [weibull_transition()] or [piecewise_exponential()] for details.
+#'
+#' @return The correlation of PFS and OS.
+#' @export
+#'
+#' @examples
+#' transition <- exponential_transition(h01 = 1.2, h02 = 1.5, h12 = 1.6)
+#' corTrans(transition)
+corTrans <- function(transition) {
+  # E(PFS) & E(OS).
+  expvalPFS <- stats::integrate(survPFS,
+    lower = 0, upper = Inf,
+    transition = transition
+  )$value
+
+  expvalOS <- stats::integrate(survOS,
+    lower = 0, upper = Inf,
+    transition = transition
+  )$value
+
+  # Var(PFS) & Var(OS).
+  expvalPFS2 <- 2 * stats::integrate(expvalPFSInteg,
+    lower = 0, upper = Inf,
+    transition = transition
+  )$value
+
+  expvalOS2 <- 2 * stats::integrate(expvalOSInteg,
+    lower = 0, upper = Inf,
+    transition = transition
+  )$value
+
+  varPFS <- expvalPFS2 - expvalPFS^2
+
+  varOS <- expvalOS2 - expvalOS^2
+
+  # E(PFS*OS).
+  expvalPFSOS <- stats::integrate(survPFSOS,
+    lower = 0, upper = Inf,
+    transition
+  )$value
+
+  # Cor(PFS, OS).
+  (expvalPFSOS - expvalPFS * expvalOS) / (varPFS * varOS)^0.5
```

Comment:
```suggestion
  (expvalPFSOS - expvalPFS * expvalOS) / sqrt(varPFS * varOS)
```
is a bit easier to read